### ISSUE TITLE
[question] Using a toolchain from a build_requires declared in a custom profile

### ISSUE THREAD

**Issue by puetzk** [conan-io/conan#6051]
### Environment Details (include every applicable attribute)
  * Operating System+version: windows
  * Conan version: 1.20.2
  * Python version: 3.7.4

### Steps to reproduce (Include if Applicable)
profile
```.ini
[settings]
os=Windows
arch=x86

[build_requires]
profile_requires/1.0@example/test
```

profile_requires/conanfile.py
```.py
from conans import ConanFile

class ProfilerequiresConan(ConanFile):
    name = "profile_requires"
    version = "1.0"
    settings = ()

    options = { 'build_requires': [False, True] }
    default_options = { 'build_requires': False }

    def requirements(self):
        if self.options.build_requires:
            self.requires('build_requires/1.0@example/test')
```

build_requires/conanfile.py
```.py
from conans import ConanFile

class BuildrequiresConan(ConanFile):
    name = "build_requires"
    version = "1.0"
    settings = "os", "arch"

    def package(self):
        # includes some utility CMake modules
        self.copy("*")

    def package_id(self):
        self.info.settings.clear()
        self.info.options.clear()

    def package_info(self):
        # uses self.settings to inject appropriate CONAN_CMAKE_TOOLCHAIN_FILE
        pass
```

consumer/conanfile.txt
```.ini
[build_requires]
build_requires/1.0@example/test

[options]
profile_requires:build_requires=True
```

### Logs (Executed commands with output) (Include/Attach if Applicable)
```sh
conan export profile_requires example/test
conan export build_requires example/test
conan install consumer --profile profile
```

> Configuration:
> [settings]
> arch=x86
> os=Windows
> [options]
> [build_requires]
> *: profile_requires/1.0
> [env]
> 
> conanfile.txt: Installing package
> Build requirements
>     profile_requires/1.0@example/test from local cache - Cache
> Build requirements packages
>     profile_requires/1.0@example/test:b0667569a18f9cc1721f72a0103ba10594bbe420 - Missing

```sh
conan editable add build_requires build_requires/1.0@example/test
conan install consumer --profile profile
```

> Configuration:
> [settings]
> arch=x86
> os=Windows
> [options]
> [build_requires]
> *: profile_requires/1.0
> [env]
> 
> ERROR: Loop detected: 'profile_requires/1.0@example/test' requires 'build_requires/1.0@example/test' which is an ancestor too

Basically, when the build_requires is in editable mode, conan appears to be acting as if the editable package itself re-declares the build_requires contained in the profile, but does so using the consumer's default_options instead of those from build_requires's recipe. This would indeed be a dependency loop build_requires -> profile_requires -> build_requires, but it's one that isn't really there (because build_requires does not set profile_requires:build_requires=True)

**Comment by puetzk** [comment]
Yes, priority:low is fine. I could easily workaround it by using conan create instead of editable. I just thought that the apparent mechanism (using the wrong set of options) deserved a report.

**Comment by lasote** [comment]
Sure, it is assigned to the 1.21 to take a look anyway so we can see how bad it looks.

**Comment by jgsogo** [comment]
Hi! It doesn't fail when it is not editable because Conan doesn't expand the graph to build `build_requires`, Conan just says its binary is not available.:

```bash
⇒  conan install conanfile.txt --profile profile
Configuration:
[settings]
arch=x86_64
os=Macos
[options]
[build_requires]
*: profile_requires/1.0@example/test
[env]

conanfile.txt: Installing package
Requirements
Packages
Build requirements
    build_requires/1.0@example/test from local cache - Cache
    profile_requires/1.0@example/test from local cache - Cache
Build requirements packages
    build_requires/1.0@example/test:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Missing
    profile_requires/1.0@example/test:b0667569a18f9cc1721f72a0103ba10594bbe420 - Missing

build_requires/1.0@example/test: WARN: Can't find a 'build_requires/1.0@example/test' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, os=Macos
- Options: 
- Dependencies: 
- Package ID: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9

ERROR: Missing prebuilt package for 'build_requires/1.0@example/test'
Try to build it from sources with "--build build_requires"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

If we try to build `build_requires`, then we get the error message about the loop:

```bash
⇒  conan install conanfile.txt --profile profile --build build_requires
Configuration:
[settings]
arch=x86_64
os=Macos
[options]
[build_requires]
*: profile_requires/1.0@example/test
[env]

build_requires/1.0@example/test: Forced build from source
ERROR: Loop detected: 'profile_requires/1.0@example/test' requires 'build_requires/1.0@example/test' which is an ancestor too
```

This is the expected behavior, Conan won't expand _build requires_ unless needed.... and Conan reports an error if there is a loop in the graph.

Note.- The default option in the recipe is overridden by the option declared in the consumer: `profile_requires:build_requires=True`.

----

The _question_ here is why the graph is expanded for the editable package. Right now Conan doesn't assume that building the editable package and providing the binaries is up to the user, instead, Conan will compute the graph to build the package too (and it detects the loop right away).

**Comment by puetzk** [comment]
Ok. I thought usually default_options set in consumer's graph wouldn't be in effect when (recursiveyl) building a dependency, except for the options of the package itself And. I *thought* `--build missing` worked fine, but you're right, it doesn't. I must have accidentally tried that in a state of minimizing the testcase were I didn't have everything set up. And it makes sense that other options would have to be propagated too, since that's could well change the package_id that needs to be built (e.g. we might have been setting `foo:shared` on a library or something like that).

Then I think that makes sense, and this package is just too weird for editable, and that's probably OK.

Then maybe this is more of a feature request/question and see if I can come up with a less strange way to do this package's job: the real goal is to he a profile for a cross-compiled platform that depends on a package containing the CMAKE_TOOLCHAIN_FILE and some supporting modules, which sets  `CONAN_CMAKE_TOOLCHAIN_FILE` in the build environment so the builtin `conans.CMake` helper will do it without needing recipe edits, but only trigger this if the package itself requests it. I.e. `profile_requires` is really named `toolchain` and `build_requires` is named `cmake-modules`, and the option is `toolchain:cmake` so the user can build a random (if well-behaved) upstream recipe (e.g. from conan-center) with `--profile x86-linux-gcc-<target> --options toolchain:cmake=True ` and we'll inject the profile and get a resulting binary built through the cross-toolchain with no recipe changes, whose package_id tags it as being for the embedded sysroot. And then I have analogous packages for toolchain:qmake, toolchain:autoconf that set up their respective environments. And between those 3 I've covered most of what I needed :-)

What this really *wants* is some way for a profile to have data files whose extracted path it can substitute in, maybe a script version of a profile analagous to the conanfile.txt -> conanfile.py relationship, where it can generate its env section like the `package_info` method can. And bonus points if it could introspect a little on the recipes it's being invoked to build (e.g. which generators are requested by the consumer) so it could make the cmake/qmake/autotools decision without needing an option :-)

**Comment by jgsogo** [comment]
First things first: we are working on using toolchains in Conan (https://github.com/conan-io/conan/pull/5919) and also implementing a new model for cross-building (https://github.com/conan-io/conan/pull/5592). These two efforts should be combined (and will) to target natively your use-case.... but it will take us some releases to get there.

----

⚠️  Danger zone: not tested and probably there is a better approach

One workaround could be to use `CONAN_CMAKE_TOOLCHAIN_FILE` environment variable, which will be translated by the CMake helper into the `-DCMAKE_TOOLCHAIN_FILE=<value>` command-line argument. 

You can have those toolchains stored in a _build_requires_ and, depending on the option (`toolchain:cmake=True`, toolchain:qmake=True`), populate the environment with a different toolchain. If the _build_requires_ is applied to every recipe (declared in the profile) it will be used to compile all of them.

This is just an idea, I haven't tested it, but I think it might work with some changes. If you feel like going deeper into this let me know, maybe it is useful mentioning it in the [cross-building section of the docs](https://docs.conan.io/en/latest/systems_cross_building/cross_building.html?highlight=cross%20build). It will also provide us with some useful working toolchains to work with for the PRs mentioned above.

**Comment by memsharded** [comment]
This was solved time ago with the conf ``tools.cmake.cmaketoolchain:user_toolchain`` conf, closing as solved.
